README.md
# LiteRT/mutex-redis

[![npm version](https://img.shields.io/npm/v/@litert/mutex-redis.svg?colorB=brightgreen)](https://www.npmjs.com/package/@litert/mutex-redis "Stable Version")
[![License](https://img.shields.io/npm/l/@litert/mutex-redis.svg?maxAge=2592000?style=plastic)](https://github.com/litert/mutex-redis/blob/master/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/litert/mutex-redis.js.svg)](https://github.com/litert/mutex-redis.js/issues)
[![GitHub Releases](https://img.shields.io/github/release/litert/mutex-redis.js.svg)](https://github.com/litert/mutex-redis.js/releases "Stable Release")

The redis driver for mutex.js.

## Requirement

- TypeScript v4.0.0 (or newer)

## Installation

```sh
npm i @litert/mutex-redis --save
```

## Document

- [简体中文版](./docs/zh-CN/README.md)

## License

This library is published under [Apache-2.0](./LICENSE) license.


CHANGES.md
# Changes Logs

## v0.2.0

- fix(redis): adapt to the new version of @litert/redis.
- config(deps): updated the deps.
- config(lint): replaced tslint with eslint.


src/lib/Common.ts
/**
 * Copyright 2020 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const DEFAULT_NAMESPACE = 'default';

export type IKeyBuilder = (namespace: string, key: string) => string;

export const DEFAULT_KEY_BUILDER: IKeyBuilder = function(
    namespace: string,
    key: string
): string {

    return `mutex:${namespace}:${key}`;
};


src/lib/Driver.Redis.ts
/**
 * Copyright 2020 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as Mutex from '@litert/mutex';
import * as Redis from '@litert/redis';
import {
    IKeyBuilder,
    DEFAULT_NAMESPACE,
    DEFAULT_KEY_BUILDER
} from './Common';

const LUA_LOCK: string = `if 1 == redis.call('setnx', KEYS[1], ARGV[1]) then

    if 1 == redis.call('pexpireat', KEYS[1], ARGV[2]) then

        return 1;
    end

    redis.call('del', KEYS[1])
end

return 0`;

const LUA_UNLOCK: string = `local val = redis.call('get', KEYS[1])

if val == ARGV[1] then

    return redis.call('del', KEYS[1])
end

return 0`;

const LUA_CHECK: string = `local val = redis.call('get', KEYS[1])

if val == ARGV[1] then

    return 1
end

return 0`;

class RedisDriver
implements Mutex.IDriver {

    private _redis: Redis.ICommandClient;

    private _ns: string;

    private _buildKey: IKeyBuilder;

    private _luaLockId!: string;

    private _luaUnlockId!: string;

    private _luaCheckId!: string;

    public constructor(
        redis: Redis.ICommandClient,
        namespace: string = DEFAULT_NAMESPACE,
        keyBuilder: IKeyBuilder = DEFAULT_KEY_BUILDER
    ) {

        this._redis = redis;
        this._ns = namespace;
        this._buildKey = keyBuilder;

        this._luaCheckId = 'cc5e3d7f05b5e6c28c556242d5a24eece2a1ae2a';
        this._luaLockId = '3a971c811a10347bbdee17970f8b6ed3cb6fd7f7';
        this._luaUnlockId = 'd03dcdcd1e51e59673c900e403199e9544f77b26';
    }

    private async _registerLuaScripts(): Promise<void> {

        this._luaLockId = await this._redis.scriptLoad(LUA_LOCK);

        this._luaUnlockId = await this._redis.scriptLoad(LUA_UNLOCK);

        this._luaCheckId = await this._redis.scriptLoad(LUA_CHECK);
    }

    public async lock(
        key: string,
        token: string,
        expiringAt: number
    ): Promise<boolean> {

        if (!expiringAt) {

            return !!await this._redis.setNX(
                this._buildKey(this._ns, key),
                token
            );
        }

        let retriable = true;

        do {

            try {

                return !!await this._redis.evalSHA(
                    this._luaLockId,
                    [ this._buildKey(this._ns, key) ],
                    [ token, expiringAt.toString() ]
                );
            }
            catch (e) {

                if (!retriable) {

                    return false;
                }

                if (e.metadata?.message?.startsWith('NOSCRIPT')) {

                    await this._registerLuaScripts();
                }

                retriable = false;
            }
        }
        while (1);

        return false;
    }

    public async unlock(
        key: string,
        token: string
    ): Promise<boolean> {

        let retriable = true;

        do {

            try {

                return !!await this._redis.evalSHA(
                    this._luaUnlockId,
                    [ this._buildKey(this._ns, key) ],
                    [ token ]
                );
            }
            catch (e) {

                if (!retriable) {

                    return false;
                }

                if (e.metadata?.message?.startsWith('NOSCRIPT')) {

                    await this._registerLuaScripts();
                }

                retriable = false;
            }
        }
        while (1);

        return false;
    }

    public async checkLocked(
        key: string,
        token: string
    ): Promise<boolean> {

        let retriable = true;

        do {

            try {

                return !!await this._redis.evalSHA(
                    this._luaCheckId,
                    [ this._buildKey(this._ns, key) ],
                    [ token ]
                );
            }
            catch (e) {

                if (!retriable) {

                    return false;
                }

                if (e.metadata?.message?.startsWith('NOSCRIPT')) {

                    await this._registerLuaScripts();
                }

                retriable = false;
            }
        }
        while (1);

        return false;
    }
}

/**
 * Create a driver for Redis.
 */
export function createRedisDriver(
    redis: Redis.ICommandClient,
    namespace: string = DEFAULT_NAMESPACE,
    keyBuilder: IKeyBuilder = DEFAULT_KEY_BUILDER
): Mutex.IDriver {

    return new RedisDriver(
        redis, namespace, keyBuilder
    );
}


src/samples/sample-01.ts
/**
 *  Copyright 2018 Angus.Fenying <fenying@litert.org>
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

// tslint:disable:no-console

import * as Mutex from "@litert/mutex";
import * as Redis from "@litert/redis";

import createRedisDriver from "../lib";

(async () => {

    const mf = Mutex.createFactory();

    mf.registerType("redis", createRedisDriver(
        await Redis.createRedisClient({
            host: "127.0.0.1",
            port: 6379
        })
    ));

    let m1 = mf.createMutex("redis", "1234");

    if (await m1.lock()) {

        console.log("[m1] Locked successfully.");

        if (!await m1.lock()) {

            console.log("[m1] Cannot relock a mutex.");
        }

        if (!await m1.lock()) {

            console.log("[m1] Cannot relock a mutex.");
        }

        if (m1.isLocked()) {

            console.log("[m1] m1 is locked.");
        }

        if (m1.isExpired()) {

            console.log("[m1] m1 is expired.");
        }

        if (await m1.unlock()) {

            console.log("[m1] Unlocked successfully.");
        }

        if (!await m1.lock()) {

            console.log("[m1] Cannot relock a mutex.");
        }

        const m2 = mf.createMutex("redis", "1234");

        if (await m2.lock()) {

            console.log("[m2] Locked successfully.");
        }
        else {

            console.log("[m2] Failed to lock.");
        }

        if (await m2.unlock()) {

            console.log("[m2] Unlocked successfully.");
        }
        else {

            console.warn("[m2] Failed to unlock.");
        }

        if (await m1.unlock()) {

            console.log("[m1] Unlocked successfully.");
        }

        if (await m1.lock()) {

            console.log("[m1] Locked successfully.");
        }

        if (await m1.unlock()) {

            console.log("[m1] Unlocked successfully.");
        }

    }

})();


src/samples/sample-02.ts
/**
 *  Copyright 2018 Angus.Fenying <fenying@litert.org>
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

// tslint:disable:no-console

import * as Mutex from "@litert/mutex";
import * as Redis from "@litert/redis";

import createRedisDriver from "../lib";

const mf = Mutex.createFactory();

let value: number = 0;

function sleep(ms: number): Promise<void> {

    return new Promise<void>((resolve) => setTimeout(resolve, ms));
}

async function executeByInterval(fn: () => Promise<any>, ms: number) {

    while (1) {

        await sleep(ms);

        await fn();
    }
}

const STARTED_AT = Date.now();

function now(): number {

    return Date.now() - STARTED_AT;
}

async function task1() {

    const mutex = mf.createMutex("default", "write-value", 1000);

    if (await mutex.lock()) {

        value = value + 1;

        console.log(JSON.stringify([now(), 1, "started"]) + ",");

        await sleep(200);

        await mutex.unlock();
    }
    else {

        console.log(JSON.stringify([now(), 1, "failed"]) + ",");
    }

    console.log(JSON.stringify([now(), 1, "ended"]) + ",");
}

async function task2() {

    const mutex = mf.createMutex("default", "write-value", 1000);

    if (await mutex.lock()) {

        value = value + 2;

        console.log(JSON.stringify([now(), 2, "started"]) + ",");

        await sleep(500);

        await mutex.unlock();
    }
    else {

        console.log(JSON.stringify([now(), 2, "failed"]) + ",");
    }

    console.log(JSON.stringify([now(), 2, "ended"]) + ",");

}

(async () => {

    mf.registerType("default", createRedisDriver(
        await Redis.createRedisClient({
            host: "127.0.0.1",
            port: 6379
        })
    ));

    executeByInterval(task1, 700);

    executeByInterval(task2, 500);

})();


src/examples/example-02.ts
/**
 * Copyright 2020 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// tslint:disable:no-console

import * as Mutex from '@litert/mutex';
import * as Redis from '@litert/redis';

import createRedisDriver from '../lib';

const mf = Mutex.createFactory();

let value: number = 0;

function sleep(ms: number): Promise<void> {

    return new Promise<void>((resolve) => setTimeout(resolve, ms));
}

async function executeByInterval(fn: () => Promise<any>, ms: number): Promise<void> {

    while (1) {

        await sleep(ms);

        await fn();
    }
}

const STARTED_AT = Date.now();

function now(): number {

    return Date.now() - STARTED_AT;
}

function createTask(step: number, interval: number): () => Promise<void> {

    return async function(): Promise<void> {

        const mutex = mf.createMutex('redis', 'write-value', 1000);

        if (await mutex.lock()) {

            value = value + step;

            console.log(JSON.stringify([now(), step, 'started']) + ',');

            await sleep(interval);

            await mutex.unlock();

            console.log(JSON.stringify([now(), step, 'ended']) + ',');
        }
        else {

            console.error(JSON.stringify([now(), step, 'failed']) + ',');
        }
    };
}

(async () => {

    const redisConn = Redis.createCommandClient({
        host: '127.0.0.1',
        port: 6379
    });

    await redisConn.connect();

    mf.registerType('redis', createRedisDriver(redisConn));

    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    executeByInterval(createTask(1, 100), 700);

    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    executeByInterval(createTask(2, 400), 500);

})().catch(console.error);


src/examples/example-01.ts
/**
 * Copyright 2020 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// tslint:disable:no-console

import * as Mutex from '@litert/mutex';
import * as Redis from '@litert/redis';

import createRedisDriver from '../lib';

async function validate(
    fn: () => Promise<boolean>,
    expected: boolean,
    whileTrue: string,
    whileFalse: string,
    whileException: string,
): Promise<boolean> {

    try {

        const expectedText = expected ? whileTrue : whileFalse;
        const unexpectedText = expected ? whileFalse : whileTrue;

        if (expected === await fn()) {

            console.log(`${expectedText} [EXPECTED]`);

            return true;
        }
        else {

            console.error(`${unexpectedText} [UNEXPECTED]`);

            return false;
        }
    }
    catch {

        console.error(`${whileException} [ERROR]`);

        return false;
    }
}

(async () => {

    const mf = Mutex.createFactory();

    const redisConn = Redis.createCommandClient({
        host: '127.0.0.1',
        port: 6379
    });

    await redisConn.connect();

    mf.registerType('redis', createRedisDriver(redisConn));

    let m1 = mf.createMutex('redis', '1234');

    if (await validate(
        async () => await m1.lock(),
        true,
        '[m1] Locked successfully.',
        '[m1] Locked failed.',
        '[m1] Error!'
    )) {

        await validate(
            async () => await m1.lock(),
            false,
            '[m1] Relocked successfully.',
            '[m1] Cannot relock a mutex.',
            '[m1] Error!'
        );

        await validate(
            async () => await m1.lock(),
            false,
            '[m1] Relocked successfully.',
            '[m1] Cannot relock a mutex.',
            '[m1] Error!'
        );

        await validate(
            async () => m1.isLocked(),
            true,
            '[m1] m1 is locked.',
            '[m1] m1 is unlocked.',
            '[m1] Error!'
        );

        await validate(
            async () => m1.isExpired(),
            false,
            '[m1] m1 is expired.',
            '[m1] m1 is locked and not expired.',
            '[m1] Error!'
        );

        await validate(
            async () => await m1.unlock(),
            true,
            '[m1] Unlocked successfully.',
            '[m1] Cannot unlock a mutex.',
            '[m1] Error!'
        );

        await validate(
            async () => await m1.lock(),
            true,
            '[m1] Locked successfully.',
            '[m1] Cannot lock a mutex.',
            '[m1] Error!'
        );

        await validate(
            async () => await m1.lock(),
            false,
            '[m1] Relocked successfully.',
            '[m1] Cannot relock a mutex.',
            '[m1] Error!'
        );

        const m2 = mf.createMutex('redis', '1234');

        await validate(
            async () => await m2.lock(),
            false,
            '[m2] Relocked successfully.',
            '[m2] Cannot relock a mutex.',
            '[m2] Error!'
        );

        await validate(
            async () => await m2.unlock(),
            false,
            '[m2] Unlocked successfully.',
            '[m2] Cannot unlock a mutex.',
            '[m2] Error!'
        );

        await validate(
            async () => await m1.unlock(),
            true,
            '[m1] Unlocked successfully.',
            '[m1] Cannot unlock a mutex.',
            '[m1] Error!'
        );

        await validate(
            async () => await m2.lock(),
            true,
            '[m2] Locked successfully.',
            '[m2] Cannot lock a mutex.',
            '[m2] Error!'
        );

        await validate(
            async () => await m1.lock(),
            false,
            '[m1] Relocked successfully.',
            '[m1] Cannot relock a mutex.',
            '[m1] Error!'
        );

        await validate(
            async () => await m1.unlock(),
            false,
            '[m1] Unlocked successfully.',
            '[m1] Cannot unlock a mutex.',
            '[m1] Error!'
        );

        await validate(
            async () => await m2.unlock(),
            true,
            '[m2] Unlocked successfully.',
            '[m2] Cannot unlock a mutex.',
            '[m2] Error!'
        );

    }

    await redisConn.close();

})().catch((e) => console.error(e));
